snaps/build-and-release-snaps-on-new-release.sh

```diff
@@ -6,147 +6,183 @@ set -eux
+      echo "Promoting mature $kube_version release to stable"
       # Promote snaps from edge to candidate
       scripts_path=$(dirname "$0")
-      version=$(get_major_minor $KUBE_VERSION)
+      version=$(get_major_minor $kube_version)
       export PROMOTE_FROM="$version/candidate"
       export PROMOTE_TO="$version/stable"
-      if [ "$BRANCH" == "latest" ]
+      if [ "$track" == "latest" ]
       then
         export PROMOTE_TO="stable $PROMOTE_TO"
       fi
       $scripts_path/promote-snaps.sh
-
-      stable_promotion='yes'
     else
-	echo "Release $KUBE_VERSION too young to promote to stable"
+	echo "Release $kube_version too young to promote to stable"
 	echo "Will promote in $(( (release_time - right_now) / 86400 )) days"
     fi
   fi
 }
 
 
-function find_release {
+function find_upstream_release_for_track {
   # Finds a suitable release for this version(stable/beta/alpha)
-  if [ "$1" == "latest" ]
+  #
+  # Input:
+  # ${TRACK} the track we are looking for (eg 1.11) or "latest"
+  #
+  # Output:
+  # ${KUBE_VERSION} the latest release version for the $TRACK passed in
+  # ${KUBE_VERSION_IS_STABLE} 'yes' if the release is considered stable upstream
+  local track="$1"
+
+  if [ "$track" == "latest" ]
   then
     url="https://dl.k8s.io/release/stable.txt"
   else
-    url="https://dl.k8s.io/release/stable-${BRANCH}.txt"
+    url="https://dl.k8s.io/release/stable-${track}.txt"
   fi
 
   export KUBE_VERSION="$(curl -s -L $url)"
-  export KUBE_VERSION_URL=$url
+  export KUBE_VERSION_IS_STABLE='yes'
 
   if [[ $KUBE_VERSION = *"Error"* ]]
   then
-    if [ "$1" == "latest" ]
-    then
-      url="https://dl.k8s.io/release/latest.txt"
-    else
-      url="https://dl.k8s.io/release/latest-${BRANCH}.txt"
-    fi
-
+    # We did not get a stable release for the track we are looking for
+    # Let's check for unstable ("latest") releases
+    url="https://dl.k8s.io/release/latest-${track}.txt"
     export KUBE_VERSION="$(curl -s -L $url)"
-    export KUBE_VERSION_URL=$url
+    export KUBE_VERSION_IS_STABLE='no'
   fi
 }
 
 
 # Main loop. Go over all versions
-for BRANCH in ${VERSIONS[@]}
+for TRACK in ${VERSIONS[@]}
 do
-  find_release "${BRANCH}"
-
-  export BRANCH
+  find_upstream_release_for_track "${TRACK}"
 
   # Work only on the available branches (eg 2.0 might not be there yet)
   if [[ $KUBE_VERSION != *"Error"* ]]
   then
     echo "Processing $KUBE_VERSION"
-    check_for_release
-    if [ "$trigger" == 'yes' ]
+    is_upstream_newer_than_edge "${TRACK}" "${KUBE_VERSION}" "${SNAP_INFO_KUBECTL}"
+    if [ "$upstream_is_newer" == 'yes' ]
     then
-      build_promote
+      build_and_promote_snaps_to_all_but_stable "${TRACK}" "${KUBE_VERSION}"
     fi
 
-    promote_stable
+    if [ "$KUBE_VERSION_IS_STABLE" == 'yes' ]
+    then
+      # Do not promote to stable something it is not stable upstream.
+      # Also do not create cdk addons for something that is not stable because it
```

Comment:
Where do we stop the creation of cdk addons? This comment seems out of place here.

Reply:
This is reference to this code https://github.com/juju-solutions/kubernetes-jenkins/blob/master/snaps/promote-snaps.sh#L48 But you are right this comment may be off in case `promote-snaps.sh` gets updated. I am removing it.